privacyidea/lib/caconnectors/msca.py

```diff
@@ -0,0 +1,293 @@
+        """
+        config[ATTR.HTTP_PROXY] = is_true(config.get(ATTR.HTTP_PROXY))
+        return config
+
+    def _connect_to_worker(self):
+        """
+        creates a connection to the middleware and returns it.
+        """
+        channel = grpc.insecure_channel('{0!s}:{1!s}'.format(self.hostname, self.port),
+                                        options=(('grpc.enable_http_proxy', int(is_true(self.http_proxy))),))
+        try:
+            grpc.channel_ready_future(channel).result(timeout=3)
+        except grpc.FutureTimeoutError:
+            log.warning("Worker seems to be offline. No connection could be established!")
+        else:
+            return CAServiceStub(channel)
+
+    @classmethod
+    def get_caconnector_description(cls):
+        """
+        Return the description of this CA connectors.
+        This contains the name as a key and the possible parameters.
+
+        :return: resolver description dict
+        :rtype:  dict
+        """
+        typ = cls.connector_type
+        config = {ATTR.HOSTNAME: 'string',
+                  ATTR.PORT: 'string',
+                  ATTR.HTTP_PROXY: 'string',
+                  ATTR.CA: 'string'}
+        return {typ: config}
+
+    def _check_attributes(self):
+        for req_key in [ATTR.HOSTNAME, ATTR.PORT]:
+            if req_key not in self.config:
+                raise CAError("required argument '{0!s}' is missing.".format(
+                    req_key))
+
+    def set_config(self, config=None):
+        self.config = config or {}
+
+        self._check_attributes()
+        self.hostname = self.config.get(ATTR.HOSTNAME)
+        self.port = self.config.get(ATTR.PORT)
+        log.error(config)
+        self.http_proxy = int(is_true(self.config.get(ATTR.HTTP_PROXY)))
+        self.templates = self.get_templates()
+        self.ca = self.config.get(ATTR.CA)
+        log.error(self.http_proxy)
+
+    def sign_request(self, csr, options=None):
+        """
+        Send a signing request to the Microsoft CA
+
+        options can be
+        WorkingDir: The directory where the configuration like openssl.cnf
+        can be found.
+        CSRDir: The directory, where to save the certificate signing
+        requests. This is relative to the WorkingDir.
+        CertificateDir: The directory where to save the certificates. This is
+        relative to the WorkingDir.
+
+        :param csr: Certificate signing request
+        :type csr: PEM string or SPKAC
+        :param options: Additional options like the validity time or the
+            template or spkac=1
+        :type options: dict
+        :return: Returns the certificate object if cert was provided instantly
+        :rtype: X509 or None
+        """
+        if self.connection:
+            reply = self.connection.SubmitCR(SubmitCRRequest(cr=csr, templateName=options.get("template"),
+                                                             caName=self.ca))
+            if reply.disposition == 3:
```

Comment:
are these dispositions defined somewhere? Should we declare some static values for this?

Reply:
We can do this in the future. The will also be some further return values.